File: alibaba-ecs/operations/create_instance.js
log.trace("Started executing example:create_instance.js flintbit")

try {
    log.trace("Inputs :: " + input)

    
    action = input.get('action')
    connector_name = input.get('connector-name')
    //access_key = input.get('access-key')
    //access_key_secret = input.get('access-key-secret')
    region = input.get('region')
    image_id = input.get('image-id')
    security_group_id = input.get('security-group-id')
    instance_type = input.get('instance-type')
    vswitch_id = input.get('vswitch-id')

    new_input = JSON.parse(input)

    connector_call_response = call.connector(connector_name)
        .set('action', action)
        //.set('access-key', access_key)
        //.set('access-key-secret', access_key_secret)
        .set('region', region)
        .set('image-id', image_id)
        .set('security-group-id', security_group_id)
        .set('instance-type', instance_type)
        .set('vswitch-id', vswitch_id)


    // Optional Fields
    if (new_input.hasOwnProperty('system-disk-category')) {
        connector_call_response.set('system-disk-category', input.get('system-disk-category'))
    }
    if (new_input.hasOwnProperty('system-disk-size')) {
        connector_call_response.set('system-disk-size', input.get('system-disk-size'))
    }
    if (new_input.hasOwnProperty('instance-charge-type')) {
        connector_call_response.set('instance-charge-type', input.get('instance-charge-type'))
    }
    if (new_input.hasOwnProperty('is-deletion-protection')) {
        connector_call_response.set('is-deletion-protection', input.get('is-deletion-protection'))
    }
    if (new_input.hasOwnProperty('instance-description')) {
        connector_call_response.set('instance-description', input.get('instance-description'))
    }
    if (new_input.hasOwnProperty('instance-password')) {
        connector_call_response.set('instance-password', input.get('instance-password'))
    }
    if (new_input.hasOwnProperty('resource-group-id')) {
        connector_call_response.set('resource-group-id', input.get('resource-group-id'))
    }
    if (new_input.hasOwnProperty('system-disk-description')) {
        connector_call_response.set('system-disk-description', input.get('system-disk-description'))
    }
    if (new_input.hasOwnProperty('system-disk-name')) {
        connector_call_response.set('system-disk-name', input.get('system-disk-name'))
    }
    if (new_input.hasOwnProperty('is-auto-renew')) {
        connector_call_response.set('is-auto-renew', input.get('is-auto-renew'))
    }
    if (new_input.hasOwnProperty('security-enhancement-strategy')) {
        connector_call_response.set('security-enhancement-strategy', input.get('security-enhancement-strategy'))
    }
    if (new_input.hasOwnProperty('auto-renew-period')) {
        connector_call_response.set('auto-renew-period', input.get('auto-renew-period'))
    }
    if (new_input.hasOwnProperty('host-name')) {
        connector_call_response.set('host-name', input.get('host-name'))
    }
    if (new_input.hasOwnProperty('io-optimized')) {
        connector_call_response.set('io-optimized', input.get('io-optimized'))
    }
    if (new_input.hasOwnProperty('tags')) {
        connector_call_response.set('tags', input.get('tags'))
    }
    if (new_input.hasOwnProperty('vlan-id')) {
        connector_call_response.set('vlan-id', input.get('vlan-id'))
    }

    connector_call_response.timeout(120000)
    call = connector_call_response.sync()

    log.trace("Connector call response : " + call)
    exit_code = call.exitcode()

    if (exit_code == 0) {
        log.trace("Create Instance Response :: \n" + call)
        // output.set('user_message', call)
        output.set('result', call.toString())
        // output.set('exit-code', 0)
    }

} catch (error) {
    log.trace("Error Message :: " + error)
    output.set('user_message', error)
    output.set('result', error)

    output.set('exit-code', -1)
}

log.trace("Finished executing example:create_instance.js flintbit")

File: openstack/operation/create_instance.rb
=begin
##########################################################################
#
#  INFIVERVE TECHNOLOGIES PTE LIMITED CONFIDENTIAL
#  __________________
# 
#  (C) INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE
#  All Rights Reserved.
#  Product / Project: Flint IT Automation Platform
#  NOTICE:  All information contained herein is, and remains
#  the property of INFIVERVE TECHNOLOGIES PTE LIMITED.
#  The intellectual and technical concepts contained
#  herein are proprietary to INFIVERVE TECHNOLOGIES PTE LIMITED.
#  Dissemination of this information or any form of reproduction of this material
#  is strictly forbidden unless prior written permission is obtained
#  from INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE.
=end

# begin

@log.trace("Started executing 'fb-cloud:openstack:operation:create_instance.rb' flintbit...")
begin
    # Flintbit Input Parameters
    @log.info("Input for create instance flintbit :: #{@input}")
    # Mandatory Input Parameters
    connector_name = @input.get('connector_name')
    action = 'create-instance'
    @target = @input.get('target')
    @username = @input.get('username')
    @password = @input.get('password')
    @port = @input.get('port')
    @protocol = @input.get('protocol')
    @version = @input.get('version')
    @domain_id = @input.get('domain-id')
    @servername = @input.get('server-name')
    @flavorId = @input.get('flavor-id')
    @imageId = @input.get('image-id')
    @networkId = @input.get('network-id')
    @project_id = @input.get('project-id')

    # optional
    request_timeout = @input.get('timeout')
    

    @log.info("Flintbit input parameters are, action : #{action} | servername : #{@servername} | flavorid : #{@flavorId} | imageid : #{@imageId} | networkid : #{@networkId}")

    @log.info('Calling openstack Connector :: ' + connector_name.to_s)
    connector_call = @call.connector(connector_name)
                          .set('action', action)
                          .set('target', @target)
                          .set('username', @username)
                          .set('password', @password)
                          .set('port', @port)
                          .set('protocol', @protocol)
                          .set('version', @version)                         
			              

    if connector_name.nil? || connector_name.empty?
        raise 'Please provide "openstack connector name (connector_name)" to create server'
    end
	
    if @project_id.nil? || @project_id.empty?
        raise 'Please provide "project id (@project_id)" to create server'
    else
        connector_call.set('project-id', @project_id)
    end

    if @domain_id.nil? || @domain_id.empty?
        raise 'Please provide "openstack domain id (@domain_id)" to create server'
    else
        connector_call.set('domain-id', @domain_id)
    end

    if @servername.nil? || @servername.empty?
        raise 'Please provide "openstack server name (server-name)" to create server'
    else
        connector_call.set('server-name', @servername)
    end

    if @flavorId.nil? || @flavorId.empty?
        raise 'Please provide "openstack flavor Id (flavor-id)" to create server'
    else
        connector_call.set('flavor-id', @flavorId)
     end

    if @imageId.nil? || @imageId.empty?
        raise 'Please provide "openstack image Id (image-id)" to create server'
    else
        connector_call.set('image-id', @imageId)
    end

    if @networkId.nil? || @networkId.empty?
        raise 'Please provide "openstack network Id (network-id)" to create server'
    else
        connector_call.set('network-id', @networkId)
        end
    if request_timeout.nil? || request_timeout.is_a?(String)
        @log.trace("Calling #{connector_name} with default timeout...")
        response = connector_call.sync
    else
        @log.trace("Calling #{connector_name} with given timeout #{request_timeout}...")
        response = connector_call.timeout(request_timeout).sync
    end

    response_exitcode = response.exitcode
    response_message = response.message

    @log.info('RESPONSE ' + response.to_s)

    serverid = response.get('serverid')

    if response_exitcode == 0
        @log.info("SUCCESS in executing #{connector_name} where, exitcode : #{response_exitcode} | message : #{response_message}")
        @output.set('exit-code', 0).set('message',response_message)

    else
        @log.error("ERROR in executing #{connector_name} where, exitcode : #{response_exitcode} | message : #{response_message}")
        @output.set('exit-code', 1).set('message', response_message)
     end

rescue Exception => e
    @log.error(e.message)
    @output.set('exit-code', 1).set('message', e.message)
end
@log.trace("Finished executing 'fb-cloud:openstack:operation:create_instance.rb' flintbit")
# end
